// ml-spectra-fitting/lib/util/selectMethod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectMethod = void 0;
const ml_levenberg_marquardt_1 = require("ml-levenberg-marquardt");
const LEVENBERG_MARQUARDT = 1;
/** Algorithm to select the method.
 * @param optimizationOptions - Optimization options
 * @returns - The algorithm and optimization options
 */
function selectMethod(optimizationOptions = {}) {
    let { kind, options } = optimizationOptions;
    kind = getKind(kind);
    switch (kind) {
        case LEVENBERG_MARQUARDT:
            return {
                algorithm: ml_levenberg_marquardt_1.levenbergMarquardt,
                optimizationOptions: checkOptions(kind, options),
            };
        default:
            throw new Error(`Unknown kind algorithm`);
    }
}
exports.selectMethod = selectMethod;
function checkOptions(kind, options = {}) {
    switch (kind) {
        case LEVENBERG_MARQUARDT:
            return Object.assign({}, lmOptions, options);
        default:
            throw new Error(`unknown kind: ${kind}`);
    }
}
function getKind(kind) {
    if (typeof kind !== 'string')
        return kind;
    switch (kind.toLowerCase().replace(/[^a-z]/g, '')) {
        case 'lm':
        case 'levenbergmarquardt':
            return LEVENBERG_MARQUARDT;
        default:
            throw new Error(`Unknown kind algorithm`);
    }
}
const lmOptions = {
    damping: 1.5,
    maxIterations: 100,
    errorTolerance: 1e-8,
};
