// nmr-processing/lib/signals/hackSignalsToXY.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hackSignalsToXY = void 0;
const couplingPatterns_1 = require("../constants/couplingPatterns");
const signalsToXY_1 = require("./signalsToXY");
/**
 * Create the xy object from an array of 1D signals.
 * @param signals Array of signals with assigned or unassigned couplings.
 */
function hackSignalsToXY(signals, options = {}) {
    let newSignals = JSON.parse(JSON.stringify(signals));
    signals.forEach((signal, s) => {
        const { js: jCouplings = [], atoms: signalAssignment = [s] } = signal;
        let { newCouplings, tempSignals } = checkCouplings(jCouplings, newSignals, signalAssignment);
        if (tempSignals.length > 0)
            newSignals.push(...tempSignals);
        newSignals[s].js = newCouplings;
        newSignals[s].atoms = signalAssignment;
    });
    return (0, signalsToXY_1.signalsToXY)(newSignals, options);
}
exports.hackSignalsToXY = hackSignalsToXY;
function checkCouplings(jCouplings, signals, signalAssignment) {
    let newSignalAssignment = signals.length - 1;
    let tempSignals = [];
    const newCouplings = jCouplings.reduce((newCouplings, jCoupling) => {
        const { atoms = [], multiplicity, coupling } = jCoupling;
        if (atoms.length === 0) {
            if (coupling && multiplicity) {
                let tempCouplings = [];
                const nbLinks = couplingPatterns_1.couplingPatterns.indexOf(multiplicity);
                for (let i = 0; i < nbLinks; i++) {
                    newSignalAssignment++;
                    tempCouplings.push({
                        coupling,
                        atoms: [newSignalAssignment],
                    });
                    tempSignals.push(formatSignal(coupling, [newSignalAssignment], signalAssignment));
                }
            }
            else {
                newCouplings.push(jCoupling);
            }
        }
        return newCouplings;
    }, []);
    return { newCouplings, tempSignals };
}
function formatSignal(coupling, newSignalAssignment, signalAssignment) {
    return {
        delta: 100000,
        atoms: newSignalAssignment,
        js: [
            {
                coupling,
                atoms: signalAssignment,
            },
        ],
    };
}
