// nmr-processing/lib/prediction/predictAllSpectra.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.predictAllSpectra = void 0;
const signalsToXY_1 = require("../signals/signalsToXY");
const predictAll_1 = require("./predictAll");
/**
 * This function will generate an object compatible with NMR-ium
 *
 * @param molecule
 * @param options
 */
async function predictAllSpectra(molecule, options = {}) {
    const predictions = await (0, predictAll_1.predictAll)(molecule, options.prediction);
    const spectra = [];
    const result = { spectra, molecules: [{ molfile: predictions.molfile }] };
    spectra.push(getProton(predictions.proton, options.simulation));
    return result;
}
exports.predictAllSpectra = predictAllSpectra;
function getProton(prediction, options) {
    const { x, y } = (0, signalsToXY_1.signalsToXY)(prediction.signals, options);
    const spectrum = {
        id: '',
        // todo Array.from is temporary for the developement
        data: { x: Array.from(x), re: Array.from(y), im: null },
        info: {
            nucleus: prediction.nucleus,
            dimension: 1,
            originFrequency: 400,
            baseFrequency: 400,
            pulseSequence: 'prediction',
            experiment: 'proton',
            isFt: true,
        },
        // todo: currently we disable the ranges because there is no ID
        //ranges: { values: prediction.ranges },
    };
    return spectrum;
}
